cirq-google/cirq_google/engine/abstract_program.py

```diff
@@ -0,0 +1,191 @@
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# coverage: ignore
+"""An interface for quantum programs.
+
+The quantum program represents a circuit (or other execution) that,
+when combined with a run context, will become a quantum job.
+"""
+
+
+import abc
+import datetime
+from typing import Dict, List, Optional, Sequence, Set, TYPE_CHECKING, Union
+import cirq
+import cirq_google.engine.client.quantum as quantum
+
+if TYPE_CHECKING:
+    import cirq_google.engine.abstract_job as abstract_job
+    import cirq_google.engine.abstract_engine as abstract_engine
+
+
+class AbstractProgram(abc.ABC):
+    """An abstract object representing a quantum program.
+
+    This program generally wraps a `Circuit` with additional metadata.
+    When combined with an appropriate RunContext, this becomes a
+    Job that can run on either an Engine service or simulator.
+    Programs can also be a batch (list of circuits) or calibration
+    requests.
+
+    This is an abstract class that inheritors should implement.
+    """
+
+    @abc.abstractmethod
+    def engine(self) -> 'abstract_engine.AbstractEngine':
+        """Returns the parent Engine object.
+
+        Returns:
+            The program's parent Engine.
+        """
+
+    @abc.abstractmethod
+    def get_job(self, job_id: str) -> 'abstract_job.AbstractJob':
+        """Returns an AbstractJob for an existing id.
+
+        Args:
+            job_id: Unique ID of the job within the parent program.
+
+        Returns:
+            A AbstractJob for this program.
+        """
+
+    @abc.abstractmethod
+    def list_jobs(
+        self,
+        created_before: Optional[Union[datetime.datetime, datetime.date]] = None,
+        created_after: Optional[Union[datetime.datetime, datetime.date]] = None,
+        has_labels: Optional[Dict[str, str]] = None,
+        execution_states: Optional[Set[quantum.enums.ExecutionStatus.State]] = None,
+    ) -> Sequence['abstract_job.AbstractJob']:
+        """Returns the list of jobs for this program.
+
+        Args:
+            project_id: A project_id of the parent Google Cloud Project.
+            program_id: Unique ID of the program within the parent project.
+            created_after: retrieve jobs that were created after this date
+                or time.
+            created_before: retrieve jobs that were created after this date
+                or time.
+            has_labels: retrieve jobs that have labels on them specified by
+                this dict. If the value is set to `*`, filters having the label
+                regardless of the label value will be filtered. For example, to
```

Comment:
Same comment on "filtered" ambiguity as above.

Reply:
Thanks for finding this one.  Changed.